resolwe/flow/tests/test_validation.py

```diff
@@ -0,0 +1,375 @@
+            {'name': 'comment', 'type': 'basic:string:'},  # implicit `required=False
+        ]
+
+        instance = {'description': 'test'}
+        with self.assertRaises(ValidationError):
+            validate_schema(instance, schema)
+
+        instance = {'value': 42}
+        validate_schema(instance, schema)
+
+        instance = {'value': 42, 'description': 'test', 'comment': 'Lorem ipsum'}
+        validate_schema(instance, schema)
+
+    def test_string_field(self):
+        schema = [
+            {'name': 'string', 'type': 'basic:string:'},
+            {'name': 'text', 'type': 'basic:text:'},
+        ]
+
+        instance = {'string': 'Test string', 'text': 'Test text'}
+        validate_schema(instance, schema)
+
+        instance = {'string': 42, 'text': 42}
+        with self.assertRaises(ValidationError):
+            validate_schema(instance, schema)
+
+    def test_boolean_field(self):
+        schema = [
+            {'name': 'true', 'type': 'basic:boolean:'},
+            {'name': 'false', 'type': 'basic:boolean:'},
+        ]
+
+        instance = {'true': True, 'false': False}
+        validate_schema(instance, schema)
+
+        instance = {'true': 'true', 'false': 'false'}
+        with self.assertRaises(ValidationError):
+            validate_schema(instance, schema)
+
+        instance = {'true': 1, 'false': 0}
+        with self.assertRaises(ValidationError):
+            validate_schema(instance, schema)
+
+        instance = {'true': 'foo', 'false': 'bar'}
+        with self.assertRaises(ValidationError):
+            validate_schema(instance, schema)
+
+    def test_integer_field(self):
+        schema = [
+            {'name': 'value', 'type': 'basic:integer:'},
+        ]
+
+        instance = {'value': 42}
+        validate_schema(instance, schema)
+
+        instance = {'value': 42.0}
+        with self.assertRaises(ValidationError):
+            validate_schema(instance, schema)
+
+        instance = {'value': 'forty-two'}
+        with self.assertRaises(ValidationError):
+            validate_schema(instance, schema)
+
+    def test_decimal_field(self):
+        schema = [
+            {'name': 'value', 'type': 'basic:decimal:'},
+        ]
+
+        instance = {'value': 42}
+        validate_schema(instance, schema)
+
+        instance = {'value': 42.0}
+        validate_schema(instance, schema)
+
+        instance = {'value': 'forty-two'}
+        with self.assertRaises(ValidationError):
+            validate_schema(instance, schema)
+
+    def test_date_field(self):
+        schema = [
+            {'name': 'date', 'type': 'basic:date:'},
+        ]
+
+        instance = {'date': '2000-12-31'}
+        validate_schema(instance, schema)
+
+        instance = {'date': '2000/01/01'}
+        validate_schema(instance, schema)
+
+        instance = {'date': '2000 04 31'}
+        validate_schema(instance, schema)
+
+        instance = {'date': '2000.06.21'}  # XXX: Is this ok?
```

Comment:
Note that one is Resolwe field `basic:date`, the other is Django's model field `models.DateTimeField`.